packages/dependencies-hierarchy/src/index.ts

```diff
@@ -223,81 +223,112 @@ function getAllDirectDependencies (projectSnapshot: ProjectSnapshot) {
   }
 }
 
+type GetTreeOpts = {
+  currentDepth: number,
+  maxDepth: number,
+  modulesDir: string,
+  includeOptionalDependencies: boolean,
+  search?: SearchFunction,
+  skipped: Set<string>,
+  registries: Registries,
+  currentPackages: PackageSnapshots,
+  wantedPackages: PackageSnapshots,
+}
+
+type DependencyInfo = { circular: boolean, dependencies: PackageNode[]}
+
 function getTree (
-  opts: {
-    currentDepth: number,
-    maxDepth: number,
-    modulesDir: string,
-    includeOptionalDependencies: boolean,
-    search?: SearchFunction,
-    skipped: Set<string>,
-    registries: Registries,
-    currentPackages: PackageSnapshots,
-    wantedPackages: PackageSnapshots,
-  },
+  opts: GetTreeOpts,
   keypath: string[],
   parentId: string,
 ): PackageNode[] {
-  if (opts.currentDepth > opts.maxDepth || !opts.currentPackages || !opts.currentPackages[parentId]) return []
+  const dependenciesCache = new Map<string, DependencyInfo>()
 
-  const deps = opts.includeOptionalDependencies === false
-    ? opts.currentPackages[parentId].dependencies
-    : {
-      ...opts.currentPackages[parentId].dependencies,
-      ...opts.currentPackages[parentId].optionalDependencies,
-    }
+  function getTreeInner (
```

Comment:
I would suggest to move this function out and pass dependenciesCache as first argument. It will reduce nesting.